From a5c722d62e2980ac2b8148a46b5c940d50fb90ce Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Fri, 12 May 2023 15:18:29 +0800
Subject: [PATCH 010/263] [Huawei] userswap: add VMA checks for register
 address

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

Add VMA checks for register address to make sure that
register address has the corresponding VMA.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/mm/userswap.c b/mm/userswap.c
index 3fcec18a354b..c9ac7e6389a7 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -500,13 +500,12 @@ bool uswap_adjust_uffd_range(struct uffdio_register *uffdio_register,
 
 	mmap_read_lock(mm);
 	vma = find_vma(mm, uffdio_register->range.start);
-	if (!vma)
+	if (!vma || vma->vm_start >= end)
 		goto out_unlock;
 	uffdio_register->range.start = vma->vm_start;
 	vma = find_vma(mm, end);
-	if (!vma)
-		goto out_unlock;
-	uffdio_register->range.len = vma->vm_end - uffdio_register->range.start;
+	if (vma && end >= vma->vm_start)
+		uffdio_register->range.len = vma->vm_end - uffdio_register->range.start;
 
 	*vm_flags |= VM_USWAP;
 
-- 
2.17.1

